feat(validator): two-phase patch validation module (MVP)

arusinova opened the pull request:
### summary

implements a validator module that mirrors capgen’s two-phase flow: **PHASE A** runs originally failing tests as a fast filter; **PHASE B** runs the full regression suite on success. adds enforced time budget, gtest xml artifacts, rollback, and posix process-group timeouts.

### changes

- validator module with two-phase orchestration (PHASE A → PHASE B)
- hard timeouts via fork/exec + process-group sigterm/sigkill
- gtest integration: --gtest_filter and --gtest_output=xml
- artifacts: ./artifacts/gtest/<phase>-<patch_id>.xml
- patch apply/restore with git restore --source=HEAD fallback to manual
- timing metrics & detailed logs, single-threaded chdir RAII guard
- lightweight xml parsing for googletest summaries

### testing
- [x]  docker build completes successfully
- [x]  did run local unit tests


### diagram

```mermaid
flowchart LR
  %% layout
  %% ---- nodes
  A([patch candidates]):::source --> B([validator]):::stage
  B --> C((two-phase flow)):::stage

  C --> PA
  subgraph PA["phase a - filter"]
    direction LR
    A1[[apply patch]]:::action --> A2[[build]]:::action --> A3[[run failing tests only]]:::action
  end

  PA --> E{tests pass?}:::decision
  E -- no --> R1["reject patch"]:::reject
  E -- yes --> PB

  subgraph PB["phase b - regression"]
    direction LR
    B1[[re-apply patch]]:::action --> B2[[build]]:::action --> B3[[run full test suite]]:::action
  end

  PB --> F{tests pass?}:::decision
  F -- no --> R2["regression detected - reject"]:::reject
  F -- yes --> P([plausible]):::result

  P --> G([validation results]):::result

  %% ---- styles
  classDef stage fill:#eef2ff,stroke:#6366f1,stroke-width:1px,rx:10,ry:10;
  classDef action fill:#ffffff,stroke:#94a3b8,rx:8,ry:8;
  classDef decision fill:#fff7ed,stroke:#f59e0b,rx:6,ry:6;
  classDef result fill:#ecfdf5,stroke:#10b981,rx:10,ry:10;
  classDef reject fill:#fef2f2,stroke:#ef4444,rx:10,ry:10;
  classDef source fill:#f0f9ff,stroke:#0ea5e9,rx:10,ry:10;

```

### known issues / limitations

- assumes single-threaded validator (due to *chdir* raii), no parallel patch validation currently
- xml parsing is a minimal line scan; not a full xml parser
- **no container isolation (!)**, builds run in repo dir and may leave artifacts
- timeouts rely on posix signals, requires a unix-like runtime (works on linux, not windows without posix layer) -> but it works for our heroku server

***

added `src/validator/notes.md` with small notes about the validator module implementation

arusinova committed:
feat(validator): implement two-phase patch validation module (MVP)

- add complete validator module with capgen-inspired two-phase validation flow
- enforce time budget (70 min) with early exit on success
- integrate gtest filtering and XML output parsing
- implement POSIX fork/exec process-group based hard timeouts
- support atomic patch application with rollback and git restore fallback
- add artifact generation for phase A + B gtest XML results
- include timing, error reporting, and build log capture
- design assumes single-threaded execution, no container isolation yet (MVP scope)

amenzies23 reviewed (approved):
Everything looks great! I just have the one question I added - but it's not a blocker.

arusinova committed:
fix: update validator path for local run - #13

arusinova committed:
refactor(prbot): remove prbot and delegate pr creation to github app

motivation:
- reduce duplication and single-source pr creation via the github app
- simplify engine responsibilities

changes:
- removed prbot module: deleted src/prbot/prbot.{h,cpp}
- removed iprbot interface from core/contracts.h
- updated orchestrator api and impl to drop pr component
- removed pr creation step from orchestrator run (delegated to app layer)
- updated main.cpp to stop constructing/injecting prbot and pr logging
- cleaned up logger: removed prbot component mapping
- updated cmake to exclude prbot and its test
- removed tests/unit/test_prbot.cpp
- updated readme to note prbot removal
- updated changelog with removal entry

impact:
- c++ engine no longer creates pull requests, github app is the sole pr creator
- pipeline_results.json remains compatible for app parsing

breaking change:
- removed iprbot and prbot; orchestrator::setcomponents signature changed

arusinova committed:
refactor(engine): remove pr-related state and json output + update readme

motivation:
- consolidate pr creation in the github app and remove redundant engine fields

changes:
- src/core/types.h: removed pr result type and pr related fields from systemstate
- src/orchestrator/orchestrator.cpp: dropped pr state initialization
- src/main.cpp: removed pr_created and pr_result from pipeline_results.json serialization
- README.md: updated architecture and output example to reflect github app handling pr

impact:
- pipeline_results.json no longer includes pr_created or pr_result
- github app flow unaffected (continues to read validation_results and patch_candidates)

arusinova committed:
feat: mvp end-to-end, and add precise inline patching

- some minor improvements in logging
- remove stub/mock logs; normalize generation logs
- pass frequency json path into mutator ctor
- switch to column-sensitive single-line replacement (not full-line replace)

arusinova committed:
feat: add pipeline_results.json

arusinova committed:
feat: remove tests dir from docker ignore, create new binary for x86_64
